Automattic/wp-calypso
# Webpack: strictly verify imports and disable nodejs polyfilling

As [discussed](https://github.com/Automattic/wp-calypso/issues/22542) in an issue earlier, the best way to make sure we keep our bundled code size lean is to make ensure we stop it from growing.  This PR focuses on that goal by modifying how we interact with the nodejs modules.

nodejs has a standard library that does not exist in browsers. this dichotomy is confusing to develop in, and in order to alleviate some of the resulting frustration webpack automatically polyfills many of these libs.  The problem: to someone that doesn't know, these libs appear as if they are just freely available but actually they don't come free, and can easily cause bloat.   We should include them with the same scrutiny that we use when including any other npm dependency.

In order to fix this going forward there were two changes I needed to make:
1. disable nodejs polyfilling within webpack
2. eslint rules so that we only depend on things explicitly declared within package.json

This could have saved us in a few cases like:
1. https://github.com/Automattic/wp-calypso/pull/17356
3. https://github.com/Automattic/wp-calypso/pull/22539

---

notes as i've been going along:
- the `jitm` data-layer uses the variable `process`.  I confirmed locally that shadowing the variable still works even though we are overwriting process with `DefinePlugin`.
- the wpcom lib uses [fs](https://github.com/Automattic/wpcom.js/blob/master/lib/site.media.js#L23) in certain conditions which is invalid in the browsers.  This is why i've added a generic mock function in its place.  Feel free to be critical of how I handled this bit because it feels like there could be a better way.
- libraries often try to access thing in the global scope for things like feature-detection.  for example [uuid](https://github.com/kelektiv/node-uuid/blob/master/lib/rng-browser.js) just calls `crypto`. I don't know why, but webpack doesn't allow for that explicitly, which is why I needed set global to `{}` with DefinePlugin
- [cachingActionCreatorFactory](https://github.com/Automattic/wp-calypso/blob/master/client/state/utils.js#L531) uses `lru-cache`, which isn't even a declared dependency in package.json.  The only reason Calypso is compiling is because it is a dependency of a dependency.  It was written for node and node for the browser. I'll replace it with `lru`.

---

**To test**
add these snippets to any file in client and make sure it crashes the build:
```js
const crypto = require('crypto');
const cert1 = new crypto.Certificate();
```

```js
import http from 'http';
http.get({
  hostname: 'localhost',
  port: 80,
  path: '/',
})
```
